CHG: [droid] Adapt to ndk15+ with unified headers

koying opened the pull request:
As the name says, those are changes to the android build chain so as to adapt to the recent NDK move to an "unified headers" approach, ie. the headers are common to all toolchain/api, vs. headers specific per arch/api as it was before.

Big advantage: large offsets are now (more-or-less) properly handled, so no need to fiddle with off_t/off64_t anymore

koying committed:
CHG: [droid] split SDK and NDK api level. They are unrelated

Use the NDK level for everything native paths related. SDK level is irrelevant.

koying committed:
ADD: [droid] Support NDK 15b+ (unified headers)

koying committed:
FIX: [droid;doc] --ndk-dir no more supported

koying committed:
FIX: [pil] use env compile options

koying committed:
CHG: [droid] default toolchain = 15c

wsnipex commented on tools/depends/configure.ac:
```diff
@@ -83,6 +83,11 @@ AC_ARG_WITH([sdk],
   [specify sdk platform version (optional for android). default is android-21])],
   [use_sdk=$withval])
 
+AC_ARG_WITH([ndk],
```
This changes behavior for an existing setting, should we go with api or ndk-version/level instead?

koying replied:
OTOH it's now symetric with the sdk one.
If you want, we can have ndk-api and sdk-api, which makes sense.

wsnipex replied:
yep, ndk-api and sdk-api sound sane

wsnipex commented on tools/depends/configure.ac:
```diff
@@ -83,6 +83,11 @@ AC_ARG_WITH([sdk],
 
+AC_ARG_WITH([ndk],
+  [AS_HELP_STRING([--with-ndk],
+  [spcify ndk level (optional for android). default is 21])],
```
specify

peak3d commented:
Can we still build with older ndk (14e for example)?

koying commented:
Yep, at least the last time I tried ;)

peak3d commented:
Ok, then I'll give it a try here

koying commented:
If by "here" you mean jenkins, that won't work as you cannot select the toolchain.

peak3d commented:
here on my private system

Memphiz commented on tools/depends/configure.ac:
```diff
@@ -524,12 +532,12 @@ if test -z $use_tarballs; then
-  if ([! test -f "$use_ndk/source.properties"] && [! test -f "$use_ndk/RELEASE.TXT"]) ; then
-    AC_MSG_ERROR("$use_ndk is not an NDK directory")
+  if ([! test -f "$use_ndk_path/source.properties"] && [! test -f "$use_ndk_path/RELEASE.TXT"]) ; then
+    AC_MSG_ERROR("$use_ndki_path is not an NDK directory")
```
Typo ndki vs ndk

Memphiz commented:
Not sure what to review exactly - those depends changes are android only or not?

peak3d commented on tools/depends/configure.ac:
```diff
@@ -524,12 +532,12 @@ if test -z $use_tarballs; then
 if test "$platform_os" == "android"; then
-  if test -z $use_ndk; then
+  if test -z $use_ndk_path; then
     AC_MSG_ERROR("NDK path is required for android")
```
would more like AC_MSG_ERROR("--with-ndk-path is required for android")

koying commented:
@Memphiz Yes, they are *supposed* to be android-only, but as they touch depends, I prefer to ping you ;)

Rechi committed:
[depends] samba-gplv3: android fixes for ndk with Unified Headers

ndk 15c has ifaddrs.h but some functions are only supported with sdk 24 or higher.
Therefore ignore ifaddrs.h from ndk and still use samba internal logic.

Memphiz commented:
Any runtime checks to carry out with this then?

koying commented:
@Memphiz Nothing specific. If it builds for ios/osx, it should be ok.
You can do a generic runtime test if you have time, but I guess a nightly post-merge would do.

Rechi reviewed (changes requested):
rename `--with-ndk` to `--with-ndk-path` is missing in docs/README.android and tools/depends/README
samba fix should be moved ahead of default toolchain bump

koying commented:
jenkins build this please

koying commented:
https://github.com/android-ndk/ndk/issues/480

I think we can close this one. GCC is really getting end-of-life in NDK.
Next step should be to move to LLVM / libc++, really, if possible.

This patch was intended for easier migration to samba 4.5.1, but it seems there is no consensus on it, so bothering with this one becomes rather pointless

MartijnKaijser commented:
you could still merge the ndk/sdk split

koying commented:
Probably, yes. Will wait to see if it's relevant to clan, though.

The PIL one is definitely something we want, though.
Currently, we build with the native python options (via sysconfig), so it's a semi-miracle it works at all ;)
e.g., from random arm32 jenkins build:
```
/usr/bin/ccache /home/jenkins/android-dev//android-toolchain-arm-21-r14b/bin/arm-linux-androideabi-gcc -DANDROID -fexceptions -fPIC -DPIC -march=armv7-a -mtune=cortex-a9 -mfloat-abi=softfp -mfpu=neon -Og -g -D_DEBUG -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/android-21 -isystem /home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include -fPIC -fno-strict-aliasing -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include -DNDEBUG -g -O3 -Wall -Wstrict-prototypes -DHAVE_LIBJPEG -DHAVE_LIBZ -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/freetype2 -I/home/jenkins/workspace/Android-ARM/tools/depends/target/pythonmodule-pil/arm-linux-androideabi-android-21-debug/libImaging -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/python2.7 -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include/python2.7 -c _imaging.c -o build/temp.linux-x86_64-2.7/_imaging.o
```

Some of the options (eg `-I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include`) come from `http://jenkins.kodi.tv/job/Android-ARM/ws/tools/depends/xbmc-depends/x86_64-linux-gnu-native/lib/python2.7/_sysconfigdata.py`

Note the **_x86_64-linux-gnu-native_** ...

Rechi commented:
@koying this needs a rebase

koying commented:
Abandonned
